[11.0][ADD] stock_account_internal_move

arkostyuk opened the pull request:
This module will allow you to configure Odoo to create journal entries when transferring goods between 2 internal locations.

Use cases:

- You have an internal location "Manufacture Stock" with all the goods that will be used for manufacturing. Goods in this location are considered "Work in Progress". On the "Manufacture Stock" location, the type is "Internal" and you should force the accounting entries to set the "Work in Progress" account.
- You have some specific locations in your warehouse where you want a separate account with the value of all the goods at that location.

Commit:
[11.0][ADD] stock_account_internal_move

etobella commented:
You should add a some information on the PR in order to have an overview of the goal of the addon.
Also, you should add a readme structure on the addon. Here you can find the template of the modules as they are expected in OCA https://github.com/OCA/maintainer-tools/tree/master/template/module

Commit:
[DOC] Add README parts

Commit:
[COV] Add tests

Commit:
[REF] Split valuation accounts on locations

max3903 commented on stock_account_internal_move/models/stock_location.py:
```diff
@@ -0,0 +1,27 @@
+    valuation_internal_account_id = fields.Many2one(
+        comodel_name='account.account',
+        string='Valuation Account (internal transfers)',
+    )
```
@arkostyuk No need for a new field. Just display the existing ones:
- Stock Valuation Account (Incoming)
- Stock Valuation Account (Outgoing)

simahawk commented on stock_account_internal_move/models/stock_location.py:
```diff
@@ -0,0 +1,27 @@
+         NOT force_accounting_entries
+         OR force_accounting_entries
+         AND valuation_internal_account_id IS NOT NULL)""",
+         'You must provide a valuation account to force accounting entries.'),
```
no translate?

simahawk commented on stock_account_internal_move/models/stock_move.py:
```diff
@@ -0,0 +1,39 @@
+                    'Could not generate a journal entry.'
+                    ' You need to either force or disable accounting on both'
+                    ' locations in order to do that.'))
+            if not location_src.force_accounting_entries:
```
elif?

arkostyuk replied:
no use, since previous `if` branch was trying to raise `ValidationError`

simahawk commented on stock_account_internal_move/tests/test_stock_move_confirmation.py:
```diff
@@ -0,0 +1,93 @@
+            'valuation_internal_account_id': self.account_from.id,
+        })
+        with self.assertRaises(ValidationError):
+            self._create_move()._action_done()
```
well done, but: no test for "_create_account_move_line"?

arkostyuk replied:
no use, since previous `if` branch was trying to raise `ValidationError`

grindtildeath commented on stock_account_internal_move/models/stock_move.py:
```diff
@@ -0,0 +1,52 @@
+            # done in _get_accounting_data_for_valuation?
+            product_valuation_accounts \
+                = move.product_id.product_tmpl_id.get_product_accounts()
+            stock_journal_id, __, __, __ \
```
`stock_journal_id = product_valuation_accounts.get('stock_journal')`

grindtildeath commented on stock_account_internal_move/models/stock_move.py:
```diff
@@ -0,0 +1,52 @@
+            if location_to.force_accounting_entries:
+                move._create_account_move_line(
+                    location_to.valuation_in_internal_account_id.id,
+                    product_valuation_accounts.get('stock_valuation').id,
```
These two lines should be switched.

grindtildeath commented on stock_account_internal_move/models/stock_move.py:
```diff
@@ -0,0 +1,52 @@
+            # created in `stock_account`
+            if not (move.location_id.usage
+                    == move.location_dest_id.usage
+                    == 'internal'):
```
I'm not a 100% sure about this check. I'll try to look into it.

grindtildeath replied:
IMO the check should simply be 
```
if not(move.location_id.force_accounting_entries or move.location_dest_id.force_accounting_entries): 
    continue
```

grindtildeath replied:
~Maybe add to the check : if product_id.product_category_id.property_valuation != 'real_time': continue~
Add to the check `if product_id.product_tmpl_id.valuation != 'real_time': continue`
And also a check `if restrict_partner_id is not null: continue`

Commit:
fixup! [REF] Split valuation accounts on locations

grindtildeath commented on stock_account_internal_move/models/stock_move.py:
```diff
@@ -0,0 +1,52 @@
+                value=move.product_id.standard_price * move.quantity_done,
+                precision_rounding=self.company_id.currency_id.rounding,
+            )
+
```
```
if location_from.force_accounting_entries and location_to.force_accounting_entries:
    move._create_account_move_line(
        location_from.valuation_out_internal_account_id.id,
        location_to.valuation_in_internal_account_id.id,
        stock_journal_id)
else:
    add the two if below here.
```

Commit:
fixup! [REF] Split valuation accounts on locations

Commit:
fixup! [REF] Split valuation accounts on locations

Commit:
[FIX] Prevent negative qtys from appearing in account.move.lines

Commit:
[IMP] Prevent valuation of products that never asked for this

max3903 commented on stock_account_internal_move/models/stock_move.py:
```diff
@@ -0,0 +1,66 @@
+        for move in res:
+            # first of all, define if we need to even valuate something
+            if move.product_id.valuation != 'real_time':
+                continue
```
@arkostyuk I don't understand this check. The accounting entry needs to be created for every costing method.

max3903 replied:
@arkostyuk You're right, sorry. I haven't had my coffee. I mixed real time and real price...

Commit:
fixup! [REF] Split valuation accounts on locations

max3903 commented on stock_account_internal_move/models/stock_move.py:
```diff
@@ -0,0 +1,66 @@
+            move.value = float_round(
+                value=move.product_id.standard_price * move.quantity_done,
+                precision_rounding=move.company_id.currency_id.rounding,
+            )
```
@arkostyuk I suggest to put the computation of the cost in a separate method that we can extend in another module.

max3903 committed:
Update test_stock_move_confirmation.py

max3903 committed:
[FIX] Duplicate field and related constraint
[IMP] Doc

grindtildeath commented:
@max3903 @arkostyuk 
There are more changes to do if you want to use the fields defined in stock_account instead of specific ones :
 - Remove the SQL constraints altogether and redefine as python constraints to be more flexible.
 - Fix the view inheritance to redefine attributes on the existing fields.

Commit:
[REF] stock_account_internal_move: Reuse fields from stock_account

* Rewrite SQL constraints to Python
* Adapt location view to display valuation accounts on internal locations

Commit:
[REF] Split internal logic to proper methods

Commit:
[COV] Update tests

Commit:
fixup! [REF] Split internal logic to proper methods

Commit:
[IMP] Respect valuation accounts on internal<>external moves

Treat corresponding valuation account counterparts on internal locations
as valuation accounts when a move is `in` or `out`, namely:

* treat location's `valuation_in_account` as a valuation on `in` moves,
* treat location's `valuation_out_account` as one on `out` moves.

Sure, the above takes place only when location has
`force_accounting_entries` on it, otherwise vanilla behavior is not to
tamper w/.

Commit:
fixup! [REF] Split internal logic to proper methods

max3903 commented:
Ping @grindtildeath @simahawk

simahawk commented on stock_account_internal_move/models/stock_location.py:
```diff
@@ -0,0 +1,45 @@
+        'usage',
+        'force_accounting_entries',
+    )
+    def _check_force_accouting_entries_internal_only(self):
```
*accounting_entries

max3903 committed:
Update stock_location.py
